### DISCUSSION TITLE
Service invoked too many times for one day: urlfetch

### DISCUSSION THREAD

**Discussion by Bedmit** [Chimildic/goofy#122] [Помощь с ошибкой] [answered]
Похоже своими экспериментами, поломал что-то важное.
Почти все скрипты, ранее работавшие без проблем, стали выдавать одну и ту же ошибку.
Например 
```
function History() {
  let History = RecentTracks.get();
  Filter.rangeDateRel(History);
  let UpdTime = 'Обновлено в ' + new Date().toTimeString().substring(0, 5);

  Playlist.saveWithReplace({
    name: 'History',
    description: UpdTime,
    public: false,
    tracks: History,  
  });
}
```
```
Exception: Service invoked too many times for one day: urlfetch.
fetch	@ Library.gs:42
fetch	@ Library.gs:3208
get	@ Library.gs:3129
getByName	@ Library.gs:1776
saveWithModify	@ Library.gs:1841
saveWithReplace	@ Library.gs:1818
History	@ History.gs:6
```

Или например тестируемый скрипт
```
function Test3() {
  let Minions = Source.getArtistsTracks({
    artist: {
    followed_include: true,},
  });

  Playlist.saveAsNew({
  name: 'Minions',
  tracks: Minions,
  });
}
```
```
Exception: Service invoked too many times for one day: urlfetch.
fetch	@ Library.gs:42
fetch	@ Library.gs:3208
get	@ Library.gs:3129
getItemsByPath	@ Library.gs:3051
(функция без названия)	@ Library.gs:315
getArtistsAlbums	@ Library.gs:313
getArtistsTracks	@ Library.gs:272
Test3	@ Test3.gs:2
```

Возможно проблема связана с `getSavedTracks`, потому что ошибки появились  после создания скрипта, создающего кэш любимых треков. Изначально я назвал его SavedTracks.json (правда потом переименовал на  LikedTracks.json).

Обновил библиотеку и права доступа. Ошибки остались. Как восстановить проект? И в чем причина ошибки?

**Comment by Chimildic** [comment, marked as answer]
Что значит почти все? По тексту ошибки, полагаю превышен лимит на количество запросов - 20 тысяч за день. Некоторые функции очень прожорливы на запросы (тот же `getArtistsTracks`).

**Reply by Bedmit** [reply]
Но часть скриптов работает. Я думал, если превышен лимит, не работает ни один.

**Reply by Chimildic** [reply]
Других документированных лимитов, связанных с `urlfetch`, я не нахожу.
Выполнение прерывается быстро или по прошествию минуты и больше? 
Часть скриптов работает по триггерам или по ручному запуску?

**Reply by Bedmit** [reply]
Ошибка появляется не сразу, а через несколько секунд. Впечатление, что процесс вычисления начинается, но при обращении к какой-то функции, прервется. 

Отрабатываются скрипты и принудительно запущенные и по триггеру. Но только те, где вычисления минимальны. Например
```
function AllHistory() {
  let AllHistory = RecentTracks.get();
  console.log('Количество треков', AllHistory.length);
}
```
Или скрипты по триггеру проверяющие соблюдение условий. Типа этого https://github.com/Chimildic/goofy/discussions/118#discussioncomment-1200331
Но только в случае, если идёт только проверка. Если по результатам проверки запускается основная часть скрипта, то он прервется ошибкой.

**Reply by Chimildic** [reply]
`RecentTracks.get` не отправляет запросов, кэш читает с диска. `Source.getSavedTracks(1)` один запрос сделает. 
В любом случае сейчас ничего не сделаем, ограничение платформы. Думаю по прошествию 24 часов с момента появления должно прийти в норму.

**Reply by Bedmit** [reply]
Только что заработали все скрипты. Может и действительно просто превышение дневного лимита. Тогда не понятно когда начинается и заканчивается отчёт этого  "дня"? 
Но главное основной вопрос видимо снят.

**Comment by whypapay** [comment]
@Chimildic а возможно как-то узнать какое количество запросов отправляет один плейлист?
такая же проблема при экспериментах и хотелось бы как-то оптимизировать свои скрипты

**Reply by Chimildic** [reply]
Наибольшее количество запросов уходит на импорт с ластфм, радио и подобному. Потому что 1 запрос это поиск 1 трека. Альтернативы этому нет из-за строения API Spotify. Есть ряд и других функций, но уже в меньшей степени влияющих на потребление. Например, повторный запрос лайков вместо выделения переменной при первом запросе. Универсальный для всего совет - это уменьшение входного количества треков. 

Подходы к оптимизации описаны в разделе [полезного опыта](https://chimildic.github.io/goofy/#/best-practices). В частности, [экономика запросов](https://chimildic.github.io/goofy/#/best-practices?id=Экономика-запросов). Там под конец говорится о функции `getCountRequest`, которая вернет количество совершенных запросов с момента запуска функции. Можете вставить логирование хоть после каждого вызова функции, чтобы увидеть что дает наибольший прирост. 
```js
console.log('Число запросов', CustomUrlFetchApp.getCountRequest());
```

**Reply by whypapay** [reply]
Спасибо!
